libraries/botbuilder-ai/src/languageTranslator.ts

```diff
@@ -29,20 +31,33 @@ export interface TranslationContext {
  * get a translated experience, and also a LUIS model allowing the user to ask to speak a language.
  */
 export class LanguageTranslator implements Middleware {
-    private luisClient: LuisClient;
     private translator: Translator;
+    private getUserLanguage: ((context: BotContext) => string) | undefined;
+    private setUserLanguage: ((context: BotContext) => Promise<boolean>) | undefined;
 
-    public constructor(translatorKey: string, protected nativeLanguages: string[], protected luisAppId: string, protected luisAccessKey: string) {
-        this.luisClient = new LuisClient();
-        this.translator = new MsTranslator({api_key: translatorKey}, true);
-        this.translator.translateArrayAsync = denodeify(this.translator, this.translator.translateArray);
+    public constructor(translatorKey: string, protected nativeLanguages: string[], noTranslatePatterns: Set<string>, getUserLanguage?: (c: BotContext) => string, setUserLanguage?: (context: BotContext) => Promise<boolean>) {
+        this.translator = new MicrosoftTranslator(translatorKey, noTranslatePatterns);
+        this.getUserLanguage = getUserLanguage;
+        this.setUserLanguage = setUserLanguage;
     }
 
     /// Incoming activity
     public async receiveActivity(context: BotContext, next: () => Promise<void>): Promise<void> {
         if (context.request.type == "message" && context.request.text) {
+            
+            if (this.setUserLanguage != undefined) {
+                let changedLanguage = await this.setUserLanguage(context);
+                if (changedLanguage) {
+                    return next();
+                }
+            }
             // determine the language we are using for this conversation
-            let sourceLanguage = this.nativeLanguages[0];
+            let sourceLanguage: string;
+            if (this.getUserLanguage != undefined) {
+                sourceLanguage = this.getUserLanguage(context);
+            } else {
+                sourceLanguage = await this.translator.detect(context.request.text);
+            }
 
             if (context.state && context.state.conversation && context.state.conversation.language) {
                 sourceLanguage = context.state.conversation.language;
```

Comment:
>sourceLanguage = context.state.conversation.language [](start = 16, length = 52)

Why call translator.detect() if you're potentially going to overwrite this?

Reply:
Are you trying to cache the results of your call here?

---
In reply to: [175600794](https://github.com/Microsoft/botbuilder-js/pull/130#discussion_r175600794) [](ancestors = 175600794)